### PR TITLE
Suggesting to add certificate information and enable SSL verification to be disabled.

### PR DESCRIPTION
**Pull request by chadgates**
In order to accommodate privately signed certificates for SSL transfer, or unknown CA's, suggest to make it possible to disable SSL certification for some partners. 
Also add some validity information to Public Certificates in order to allow management of certificates. Potentially a some clean up script could inform administrators if there are soon-to-expire certificates to prevent outage due to expired certificates.

### CODE REVIEW

**Comment by abhishek-ram** [comment]
Please add migrations and make sure that tests work..

File: pyas2/models.py [hunk 1]
```diff
@@ -19,6 +19,10 @@
 from pyas2.utils import run_post_send
 from pyas2.utils import store_file
 
+from pyas2lib.utils import pem_to_der
+from OpenSSL import crypto
```
Comment (abhishek-ram) [inline pyas2/models.py]: Please add pyopenssl to the requirements and setup file

File: pyas2/models.py [hunk 2]
```diff
@@ -39,6 +43,36 @@ class PublicCertificate(models.Model):
         verbose_name=_('Verify Certificate'), default=True,
         help_text=_('Uncheck this option to disable certificate verification.'))
 
+    certificate_valid_from = models.DateTimeField(null=True, blank=True)
```
Comment (abhishek-ram) [inline pyas2/models.py]: rename the fields as valid_from and valid_to

File: pyas2/models.py [hunk 3]
```diff
@@ -19,6 +19,10 @@
 from pyas2.utils import run_post_send
 from pyas2.utils import store_file
 
+from pyas2lib.utils import pem_to_der
+from OpenSSL import crypto
+from datetime import datetime
+
 
 class PrivateKey(models.Model):
```
Comment (abhishek-ram) [inline pyas2/models.py]: Can we extract the info for private keys as well?

File: pyas2/models.py [hunk 4]
```diff
@@ -39,6 +43,36 @@ class PublicCertificate(models.Model):
         verbose_name=_('Verify Certificate'), default=True,
         help_text=_('Uncheck this option to disable certificate verification.'))
 
+    certificate_valid_from = models.DateTimeField(null=True, blank=True)
+    certificate_valid_to = models.DateTimeField(null=True, blank=True)
+
+    def extract_info(self):
```
Comment (abhishek-ram) [inline pyas2/models.py]: Also it would be good to save the certificate serial number.

File: pyas2/models.py [hunk 5]
```diff
@@ -39,6 +43,36 @@ class PublicCertificate(models.Model):
...
+        '''
+
+        der = pem_to_der(self.certificate)
+        cert = crypto.load_certificate(crypto.FILETYPE_ASN1, der[0])
+
+        cert_info = {'valid_from': None,
+                     'valid_to': None}
+
+        if der:
+            cert_info['valid_from'] = datetime.strptime(cert.get_notBefore().decode('utf8'), "%Y%m%d%H%M%SZ")
```
Comment (abhishek-ram) [inline pyas2/models.py]: I think you can also use the django.utils.dateparse module for doing this.

Comment (chadgates) [reply pyas2/models.py]: The output of get_notBefore is: 
b'20190301181722Z' (or decoded: '20190301181722Z') 
for django.utils.dateparse to do correctly, the value would have to be: 
'2019-03-01 18:17:22Z'

File: pyas2/models.py (hunk 4 above)
Comment (chadgates) [reply pyas2/models.py]: Yes, can be done too. Not sure what field type to use for this though. BinaryField, FloatField ? Seems the serial would be a max 20 byte value.

File: pyas2/models.py (hunk 2 above)
Comment (chadgates) [reply pyas2/models.py]: done.

File: pyas2/models.py (hunk 1 above)
Comment (chadgates) [reply pyas2/models.py]: Added, however, that library already installed through pyas2lib.

File: pyas2/models.py (hunk 4 above)
Comment (abhishek-ram) [reply pyas2/models.py]: use CharField

File: pyas2/models.py (hunk 5 above)
Comment (abhishek-ram) [reply pyas2/models.py]: Okay

File: pyas2/models.py [hunk 6]
```diff
@@ -376,7 +429,7 @@ def send_message(self, header, payload):
         # Send the message to the partner
         try:
             response = requests.post(
-                self.partner.target_url, auth=auth, headers=header, data=payload)
+                self.partner.target_url, auth=auth, headers=header, data=payload, verify=self.partner.https_verify_ssl)
```
Comment (abhishek-ram) [inline pyas2/models.py]: what line lengths are you using, max will be 100

File: pyas2/models.py (hunk 3 above)
Comment (abhishek-ram) [reply pyas2/models.py]: Okay move this to utils and add test case for this function, one for private and another for public key...

Comment (chadgates) [reply pyas2/models.py]: Moved, please check PR on pyas2-lib. New release of pyas2-lib required for this to be finalized.

File: pyas2/models.py (hunk 4 above)
Comment (chadgates) [reply pyas2/models.py]: OK to do: 
serial_number = models.CharField(max_length=64, null=True, blank=True) 
?

Comment (abhishek-ram) [reply pyas2/models.py]: lets call it serial_number

File: pyas2/models.py (hunk 6 above)
Comment (chadgates) [reply pyas2/models.py]: 120 was mine, changed to 100 and have this in next commit.

File: pyas2/models.py (hunk 3 above)
Comment (abhishek-ram) [reply pyas2/models.py]: I have released version 1.1.1, please update so tests will pass

File: pyas2/migrations/0002_auto_20190531_1954.py [hunk 7]
```diff
@@ -0,0 +1,23 @@
+# Generated by Django 2.2.1 on 2019-05-31 19:54
```
Comment (abhishek-ram) [inline pyas2/migrations/0002_auto_20190531_1954.py]: Create only one migration file for this change please.

Comment (abhishek-ram) [reply pyas2/migrations/0002_auto_20190531_1954.py]: @chadgates please take care of the above comment and I will merge

Comment (chadgates) [reply pyas2/migrations/0002_auto_20190531_1954.py]: done

**Pull request merged**